Code/Tools/AssetProcessor/native/AssetManager/SourceAssetReference.cpp

```diff
@@ -0,0 +1,108 @@
+/*
+ * Copyright (c) Contributors to the Open 3D Engine Project.
+ * For complete copyright and license terms please see the LICENSE at the root of this distribution.
+ *
+ * SPDX-License-Identifier: Apache-2.0 OR MIT
+ *
+ */
+
+#include <native/AssetManager/SourceAssetReference.h>
+
+namespace AssetProcessor
+{
+    SourceAssetReference::SourceAssetReference(AZ::IO::PathView absolutePath)
+    {
+        IPathConversion* pathConversion = AZ::Interface<IPathConversion>::Get();
+
+        AZ_Assert(pathConversion, "IPathConversion interface is not available");
+
+        QString relativePath, scanfolderPath;
+        pathConversion->ConvertToRelativePath(absolutePath.FixedMaxPathStringAsPosix().c_str(), relativePath, scanfolderPath);
+        auto* scanFolderInfo = pathConversion->GetScanFolderForFile(scanfolderPath);
+
+        AZ_Assert(scanFolderInfo, "Failed to find scanfolder for " AZ_STRING_FORMAT, AZ_STRING_ARG(absolutePath.Native()));
+
+        m_scanfolderId = scanFolderInfo->ScanFolderID();
+        m_absolutePath = absolutePath;
+        m_relativePath = relativePath.toUtf8().constData();
+        m_scanfolderPath = scanfolderPath.toUtf8().constData();
+
+        Normalize();
+    }
+
+    SourceAssetReference::SourceAssetReference(AZ::IO::PathView scanFolderPath, AZ::IO::PathView pathRelativeToScanFolder)
+    {
+        AZ_Assert(!scanFolderPath.Native().empty(), "scanfolderPath is empty");
+        AZ_Assert(!pathRelativeToScanFolder.Native().empty(), "pathRelativeToScanFolder is empty");
+
+        m_scanfolderPath = scanFolderPath;
+        m_relativePath = pathRelativeToScanFolder;
+        m_absolutePath = m_scanfolderPath / m_relativePath;
+
+        IPathConversion* pathConversion = AZ::Interface<IPathConversion>::Get();
+
+        AZ_Assert(pathConversion, "IPathConversion interface is not available");
+
+        auto* scanFolderInfo = pathConversion->GetScanFolderForFile(m_scanfolderPath.c_str());
+
+        AZ_Assert(scanFolderInfo, "Failed to find scanfolder for " AZ_STRING_FORMAT, AZ_STRING_ARG(m_absolutePath.Native()));
+
+        m_scanfolderId = scanFolderInfo->ScanFolderID();
+
+        Normalize();
+    }
+
+    bool SourceAssetReference::operator==(const SourceAssetReference& other) const
+    {
+        return m_absolutePath == other.m_absolutePath;
+    }
+
+    bool SourceAssetReference::operator!=(const SourceAssetReference& other) const
+    {
+        return !operator==(other);
+    }
+
+    bool SourceAssetReference::operator<(const SourceAssetReference& other) const
+    {
+        return m_absolutePath < other.m_absolutePath;
+    }
+
+    SourceAssetReference::operator bool() const
+    {
+        return !m_absolutePath.empty();
```

Comment:
Can the source asset can be uniquely identified via its scan folder and relative path as well? If that's the case, maybe this function should return true if !m_relativePath.empty() && !m_scanfolderPath.empty()

Reply:
It can, but SourceAssetReference will always resolve the absolute path, since that's really what its meant to use for referencing files